In [121]:
import numpy as np    # linear algebra
import pandas as pd   # data processing


import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/data-science-circle-challenge/test.csv
/kaggle/input/data-science-circle-challenge/train.csv


In [122]:
import matplotlib.pyplot as plt #plotting
import seaborn as sns           #plotting
%matplotlib inline

# Getting the data

In [123]:
df=pd.read_csv('/kaggle/input/data-science-circle-challenge/train.csv')

In [124]:
df.shape

(395661, 434)

In [125]:
#This loop determines the columns with nan values more than 25% so we can drop them  
i=0.25*395661
lst=['TransactionID']
for col in df.columns:
    if sum(df[col].isnull())>i:
        lst.append(col)

In [126]:
df=df.drop(columns=lst)

In [127]:
df=df.reset_index()            #resetting the index
df=df.drop(columns=['index'])

In [128]:
pd.set_option('display.max_columns', None)   #to show all columns

# Now let's divide the data so we can deal with it easily

In [129]:
df_cat=df.loc[:,['ProductCD','card4','card6','P_emaildomain']] #the categorical columns

In [130]:
df_cat.head()

ProductCD card4   card6  P_emaildomain
0         W  visa   debit  anonymous.com
1         C  visa  credit    hotmail.com
2         W  visa   debit      gmail.com
3         W  visa   debit      gmail.com
4         W  visa   debit            NaN

In [131]:
#first group
df_g1=df.loc[:,['isFraud','TransactionDT','TransactionAmt','card1','card2','card3','card5','addr1','addr2']]

In [132]:
#second group
df_g2=df.iloc[:,13:70]

In [133]:
#third group
df_g3=df.iloc[:,70:125]

In [134]:
#fourth group
df_g4=df.iloc[:,125:]

In [135]:
from sklearn.impute import SimpleImputer  #to deal with nan values

In [136]:
df_1=SimpleImputer(strategy='median').fit_transform(df_g1)   #filling with median

In [137]:
df_1=pd.DataFrame(df_1,columns=df_g1.columns,index=df_g1.index)

In [138]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395661 entries, 0 to 395660
Data columns (total 9 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   isFraud         395661 non-null  float64
 1   TransactionDT   395661 non-null  float64
 2   TransactionAmt  395661 non-null  float64
 3   card1           395661 non-null  float64
 4   card2           395661 non-null  float64
 5   card3           395661 non-null  float64
 6   card5           395661 non-null  float64
 7   addr1           395661 non-null  float64
 8   addr2           395661 non-null  float64
dtypes: float64(9)
memory usage: 27.2 MB


In [139]:
df_g2.shape #a lot of data so we divide it again

(395661, 57)

In [140]:
#section 1
df_g2_1=df_g2.iloc[:,:17]

In [141]:
df_g2_1=df_g2_1.fillna(0)   #the median of most coulmns is zero so we fill with 0

In [142]:
#section 2
df_g2_d2=df_g2.iloc[:,17:38]

In [143]:
df_g2_d2.head()

V12  V13  V14  V15  V16  V17  V18  V19  V20  V21  V22  V23  V24  V25  V26  \
0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0   
1  0.0  0.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
2  2.0  2.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0   
3  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0   
4  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0   

   V27  V28  V29  V30  V31  V32  
0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  1.0  1.0  
2  0.0  0.0  1.0  1.0  0.0  0.0  
3  0.0  0.0  1.0  1.0  0.0  0.0  
4  0.0  0.0  1.0  1.0  0.0  0.0

In [144]:
df_g2_2=SimpleImputer(strategy='median').fit_transform(df_g2_d2) #we fill with median

In [145]:
df_g2_2=pd.DataFrame(df_g2_2,columns=df_g2_d2.columns,index=df_g2_d2.index)

In [146]:
#section 3
df_g2_d3=df_g2.iloc[:,38:]

In [147]:
df_g2_d3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395661 entries, 0 to 395660
Data columns (total 19 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   V33     344760 non-null  float64
 1   V34     344760 non-null  float64
 2   V53     344100 non-null  float64
 3   V54     344100 non-null  float64
 4   V55     344100 non-null  float64
 5   V56     344100 non-null  float64
 6   V57     344100 non-null  float64
 7   V58     344100 non-null  float64
 8   V59     344100 non-null  float64
 9   V60     344100 non-null  float64
 10  V61     344100 non-null  float64
 11  V62     344100 non-null  float64
 12  V63     344100 non-null  float64
 13  V64     344100 non-null  float64
 14  V65     344100 non-null  float64
 15  V66     344100 non-null  float64
 16  V67     344100 non-null  float64
 17  V68     344100 non-null  float64
 18  V69     344100 non-null  float64
dtypes: float64(19)
memory usage: 57.4 MB


In [148]:
df_g2_3=SimpleImputer(strategy='median').fit_transform(df_g2_d3)   #filling with median

In [149]:
df_g2_3=pd.DataFrame(df_g2_3,columns=df_g2_d3.columns,index=df_g2_d3.index)

In [150]:
df_g3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395661 entries, 0 to 395660
Data columns (total 55 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   V70     344100 non-null  float64
 1   V71     344100 non-null  float64
 2   V72     344100 non-null  float64
 3   V73     344100 non-null  float64
 4   V74     344100 non-null  float64
 5   V75     335937 non-null  float64
 6   V76     335937 non-null  float64
 7   V77     335937 non-null  float64
 8   V78     335937 non-null  float64
 9   V79     335937 non-null  float64
 10  V80     335937 non-null  float64
 11  V81     335937 non-null  float64
 12  V82     335937 non-null  float64
 13  V83     335937 non-null  float64
 14  V84     335937 non-null  float64
 15  V85     335937 non-null  float64
 16  V86     335937 non-null  float64
 17  V87     335937 non-null  float64
 18  V88     335937 non-null  float64
 19  V89     335937 non-null  float64
 20  V90     335937 non-null  float64
 21  V91     33

## It seems that dividing group 2 isn't helpful specially when it's all V columns

In [151]:
df_3=SimpleImputer(strategy='median').fit_transform(df_g3)   #filling the data with median

In [152]:
df_3=pd.DataFrame(df_3,columns=df_g3.columns,index=df_g3.index)

In [153]:
df_g4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395661 entries, 0 to 395660
Data columns (total 56 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   V125    395454 non-null  float64
 1   V126    395454 non-null  float64
 2   V127    395454 non-null  float64
 3   V128    395454 non-null  float64
 4   V129    395454 non-null  float64
 5   V130    395454 non-null  float64
 6   V131    395454 non-null  float64
 7   V132    395454 non-null  float64
 8   V133    395454 non-null  float64
 9   V134    395454 non-null  float64
 10  V135    395454 non-null  float64
 11  V136    395454 non-null  float64
 12  V137    395454 non-null  float64
 13  V279    395652 non-null  float64
 14  V280    395652 non-null  float64
 15  V281    394819 non-null  float64
 16  V282    394819 non-null  float64
 17  V283    394819 non-null  float64
 18  V284    395652 non-null  float64
 19  V285    395652 non-null  float64
 20  V286    395652 non-null  float64
 21  V287    39

In [154]:
df_g4.shape

(395661, 56)

In [155]:
df_4=SimpleImputer(strategy='median').fit_transform(df_g4) #filling with median

In [156]:
df_4=pd.DataFrame(df_4,columns=df_g4.columns,index=df_g4.index)

In [157]:
df_cat.head()

ProductCD card4   card6  P_emaildomain
0         W  visa   debit  anonymous.com
1         C  visa  credit    hotmail.com
2         W  visa   debit      gmail.com
3         W  visa   debit      gmail.com
4         W  visa   debit            NaN

In [158]:
for col in df_cat.columns:
    df_cat[col]=df_cat[col].astype('category')

In [159]:
df_cat.P_emaildomain.value_counts()

gmail.com           153115
yahoo.com            67792
hotmail.com          30427
anonymous.com        24755
aol.com              18936
comcast.net           5278
icloud.com            4174
outlook.com           3397
msn.com               2741
att.net               2693
live.com              2020
sbcglobal.net         2011
verizon.net           1807
ymail.com             1574
bellsouth.net         1266
me.com                1039
yahoo.com.mx          1029
cox.net                966
optonline.net          657
charter.net            558
live.com.mx            500
rocketmail.com         434
mail.com               377
earthlink.net          341
gmail                  330
mac.com                305
outlook.es             292
juno.com               218
hotmail.fr             208
aim.com                208
hotmail.es             204
windstream.net         203
roadrunner.com         197
frontier.com           189
embarqmail.com         171
twc.com                164
web.de                 162
n

In [160]:
df_cat.P_emaildomain=df_cat.P_emaildomain.replace('gmail','gmail.com') 
#converting the gmail to .com as usual

In [161]:
df_cat.P_emaildomain=df_cat.P_emaildomain.replace(['roadrunner.com','frontier.com','embarqmail.com',
                    'twc.com','web.de','netzero.com','prodigy.net.mx','q.com','centurylink.net', 
                'frontiernet.net','cfl.rr.com','netzero.net','suddenlink.net','sc.rr.com',
                              'cableone.net','gmx.de','yahoo.fr','yahoo.es','hotmail.co.uk',
                'protonmail.com','yahoo.de','ptd.net','live.fr','yahoo.co.uk','hotmail.de',
                              'yahoo.co.jp','servicios-ta.com'],'others')
#replacing the unusual domains with 'others'

In [162]:
df_cat.P_emaildomain.value_counts()

gmail.com         153445
yahoo.com          67792
hotmail.com        30427
anonymous.com      24755
aol.com            18936
comcast.net         5278
icloud.com          4174
outlook.com         3397
others              2776
msn.com             2741
att.net             2693
live.com            2020
sbcglobal.net       2011
verizon.net         1807
ymail.com           1574
bellsouth.net       1266
me.com              1039
yahoo.com.mx        1029
cox.net              966
optonline.net        657
charter.net          558
live.com.mx          500
rocketmail.com       434
mail.com             377
earthlink.net        341
mac.com              305
outlook.es           292
juno.com             218
hotmail.fr           208
aim.com              208
hotmail.es           204
windstream.net       203
Name: P_emaildomain, dtype: int64

In [163]:
df_0=SimpleImputer(strategy='most_frequent').fit_transform(df_cat)  #filling with mode

In [164]:
df_0=pd.DataFrame(df_0,columns=df_cat.columns,index=df_cat.index)

In [165]:
for col in df_0.columns:
    df_0[col]=df_0[col].astype('category')    #make the columns types as category to encode them

In [166]:
for col in df_0.columns:
    df_0[col]=df_0[col].cat.codes         #encoding the columns

In [167]:
visa=pd.concat([df_0,df_1,df_g2_1,df_g2_2,df_g2_3,df_3,df_4],axis=1)  #our clean dataset

In [168]:
visa.head()

ProductCD  card4  card6  P_emaildomain  isFraud  TransactionDT  \
0          4      3      2              1      0.0      5261726.0   
1          0      3      1             10      0.0     10374074.0   
2          4      3      2              9      0.0      2397517.0   
3          4      3      2              9      0.0      6150927.0   
4          4      3      2              9      0.0     10025230.0   

   TransactionAmt    card1  card2  card3  card5  addr1  addr2   C1   C2   C3  \
0          59.000  17412.0  451.0  150.0  226.0  264.0   87.0  1.0  1.0  0.0   
1          64.347   7949.0  142.0  185.0  226.0  299.0   87.0  1.0  1.0  0.0   
2        1308.500   9500.0  321.0  150.0  226.0  251.0   87.0  1.0  1.0  0.0   
3          57.950  10112.0  360.0  150.0  166.0  264.0   87.0  1.0  1.0  0.0   
4          53.950   9500.0  321.0  150.0  226.0  272.0   87.0  1.0  1.0  0.0   

    C4   C5   C6   C7   C8   C9  C10  C11  C12  C13  C14    D1   D10   D15  \
0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  4.0  1.0  42.0  42.0  42.0   
1  1.0  0.0  1.0  1.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0   0.0   0.0   0.0   
2  0.0  1.0  1.0  0.0  0.0  2.0  0.0  1.0  0.0  1.0  1.0   4.0   5.0   5.0   
3  0.0  1.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  3.0  1.0  14.0  13.0  13.0   
4  0.0  1.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  1.0  1.0   0.0   0.0   0.0   

   V12  V13  V14  V15  V16  V17  V18  V19  V20  V21  V22  V23  V24  V25  V26  \
0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0   
1  0.0  0.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
2  2.0  2.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0   
3  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0   
4  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0   

   V27  V28  V29  V30  V31  V32  V33  V34  V53  V54  V55  V56  V57  V58  V59  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  1.0   
2  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  2.0  2.0  1.0  1.0  0.0  0.0  0.0   
3  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0   
4  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0   

   V60  V61  V62  V63  V64  V65  V66  V67  V68  V69  V70  V71  V72  V73  V74  \
0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0   
2  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0   

   V75  V76  V77  V78  V79  V80  V81  V82  V83  V84  V85  V86  V87  V88  V89  \
0  1.0  1.0  1.0  1.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  0.0   
1  0.0  0.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  0.0   
2  2.0  2.0  1.0  1.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  0.0   
3  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  0.0   
4  1.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  0.0   

   V90  V91  V92  V93  V94  V95  V96  V97  V98  V99  V100  V101  V102  V103  \
0  0.0  0.0  0.0  0.0  0.0  0.0  2.0  0.0  0.0  2.0   0.0   0.0   0.0   0.0   
1  0.0  0.0  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0   0.0   0.0   0.0   0.0   
2  1.0  1.0  0.0  0.0  0.0  0.0  2.0  2.0  0.0  0.0   0.0   0.0   0.0   0.0   
3  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0   0.0   0.0   0.0   
4  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0   0.0   0.0   0.0   

   V104  V105  V106  V107  V108  V109  V110  V111  V112  V113  V114  V115  \
0   0.0   0.0   0.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
1   0.0   0.0   0.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2   0.0 

In [169]:
visa.shape

(395661, 181)

# Training and testing the data

In [170]:
from sklearn.model_selection import StratifiedShuffleSplit  
#to make the train and test sets stratified

In [171]:
split=StratifiedShuffleSplit(n_splits=1,test_size=0.2,random_state=42)

In [172]:
for train_index,test_index in split.split(visa,visa['isFraud']):
    train_set=visa.loc[train_index]
    test_set=visa.loc[test_index]

In [173]:
x_train=train_set.drop(columns=['isFraud'])
y_train=train_set['isFraud']

In [174]:
x_test=test_set.drop(columns=['isFraud'])
y_test=test_set['isFraud']

In [175]:
from sklearn.linear_model import SGDClassifier       
sgd=SGDClassifier(max_iter=1000,random_state=42)   

In [176]:
from sklearn.model_selection import cross_val_score
cross_val_score(sgd, x_train, y_train, cv=5, scoring="roc_auc")  #try using the sgd model
#and it's so bad

array([0.47921757, 0.48843972, 0.48443895, 0.47552821, 0.475525  ])

In [177]:
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import confusion_matrix

In [178]:
from sklearn.feature_selection import SelectKBest   #for feature selection
from sklearn.feature_selection import f_classif

In [179]:
from sklearn.preprocessing import MinMaxScaler     #for normalization
scaler = MinMaxScaler(feature_range=(0, 1))
x_train_norm = scaler.fit_transform(x_train)
x_train_norm=pd.DataFrame(x_train_norm,columns=x_train.columns,index=x_train.index)

In [180]:
#We divide the data so we can make feature selection in less data

x_train_1=x_train_norm.iloc[:,:40]
x_train_2=x_train_norm.iloc[:,40:80]
x_train_3=x_train_norm.iloc[:,80:120]
x_train_4=x_train_norm.iloc[:,120:]

In [181]:
test = SelectKBest(score_func=f_classif, k=20)
fit = test.fit(x_train_1, y_train)
print(sorted(zip(fit.scores_,x_train_1.columns),reverse=True))

[(8011.4538344212215, 'V18'), (8009.365720021718, 'V17'), (7753.878179702135, 'ProductCD'), (7635.271656048067, 'V15'), (7536.758662260059, 'card3'), (7398.09009545735, 'V16'), (6426.080964288048, 'V21'), (5879.598344962972, 'V22'), (3254.30430562587, 'card6'), (1764.2217813391856, 'D15'), (1629.3004660715135, 'D10'), (1451.5325002796812, 'D1'), (479.89353874309273, 'V20'), (477.18633164031615, 'C2'), (361.9373419555743, 'C8'), (357.7374576390533, 'C12'), (355.433904728606, 'V12'), (349.7371061450969, 'card5'), (329.2195159290234, 'C1'), (319.9574196002965, 'C4'), (319.4235250167152, 'C9'), (317.8460988040677, 'V19'), (300.53969606678913, 'C5'), (282.3583222904308, 'C7'), (281.6946193227385, 'V13'), (280.8977406422215, 'C10'), (264.46350159474184, 'C11'), (155.43144537639571, 'C6'), (120.22127361398587, 'addr2'), (117.30125980612549, 'P_emaildomain'), (63.69098454321798, 'card1'), (53.816895860142345, 'TransactionDT'), (45.70525255113368, 'addr1'), (44.233786653165986, 'TransactionAmt'

In [182]:
test = SelectKBest(score_func=f_classif, k=20)
fit = test.fit(x_train_2, y_train)
print(sorted(zip(fit.scores_,x_train_2.columns),reverse=True))

[(8874.913205339339, 'V79'), (8023.000988239373, 'V33'), (7729.795211529822, 'V34'), (7679.50760326472, 'V74'), (7501.612678618447, 'V80'), (6690.395561327104, 'V73'), (6647.466463228551, 'V58'), (6291.383756935072, 'V31'), (6272.966671157327, 'V57'), (6129.042802589443, 'V32'), (5815.417111229834, 'V77'), (5623.673693629137, 'V72'), (5530.447960627049, 'V23'), (5364.142762835732, 'V78'), (5287.397911020106, 'V71'), (5076.266166984329, 'V63'), (4984.493404648934, 'V59'), (4894.2522168901005, 'V60'), (4866.974863832403, 'V64'), (3685.072087385198, 'V24'), (3356.6913560376647, 'V29'), (3342.429005825307, 'V69'), (3132.9958938477084, 'V70'), (3108.132068217774, 'V30'), (1492.235154310643, 'V67'), (1431.2830209768063, 'V62'), (887.0646910890134, 'V56'), (788.2281409622663, 'V61'), (533.5011295409856, 'V55'), (403.4267229154975, 'V66'), (326.8413722716204, 'V75'), (272.4019468995149, 'V76'), (181.87818103261773, 'V53'), (100.29193117065357, 'V54'), (54.47791342365108, 'V26'), (7.62185339166

In [183]:
test = SelectKBest(score_func=f_classif, k=20)
fit = test.fit(x_train_3, y_train)
print(sorted(zip(fit.scores_,x_train_3.columns),reverse=True))

[(16109.825120029476, 'V86'), (16077.84487235633, 'V87'), (8177.887575006343, 'V94'), (7769.484149049315, 'V81'), (7561.312276199769, 'V93'), (7408.210056355023, 'V92'), (6518.676186717063, 'V85'), (6507.444107757313, 'V84'), (4632.971001133569, 'V111'), (4213.455149701411, 'V113'), (3473.5985036858006, 'V112'), (3344.3384293021563, 'V90'), (3085.4412987833985, 'V91'), (2902.2114350095017, 'V108'), (2711.473554186116, 'V114'), (2100.0266363978453, 'V110'), (1959.4821913799253, 'V116'), (1257.6189615969379, 'V83'), (1141.9057448987762, 'V109'), (954.2528023753079, 'V115'), (728.4931102810682, 'V82'), (686.1763766754281, 'V117'), (672.7655253831612, 'V119'), (466.18970276633576, 'V118'), (273.2827341345857, 'V98'), (116.08463977455875, 'V120'), (50.855247736866076, 'V100'), (50.42720201529457, 'V99'), (8.89610375299693, 'V96'), (7.742914637866098, 'V103'), (7.712593884949361, 'V89'), (6.8897282766231465, 'V102'), (6.527340622713781, 'V101'), (5.854983877727407, 'V97'), (4.89061752881376,

In [184]:
test = SelectKBest(score_func=f_classif, k=20)
fit = test.fit(x_train_4, y_train)
print(sorted(zip(fit.scores_,x_train_4.columns),reverse=True))

[(6141.561859102176, 'V123'), (5575.978074432698, 'V302'), (5378.381052442229, 'V304'), (4852.446661530601, 'V303'), (4046.8439735919005, 'V125'), (3926.9219286310145, 'V283'), (3464.177309983725, 'V282'), (1813.8376062327088, 'V124'), (1739.840765836295, 'V281'), (722.3261724396314, 'V315'), (579.0960788582132, 'V289'), (569.8315356122171, 'V313'), (545.3212670444109, 'V290'), (491.66818861219593, 'V314'), (442.0188313919757, 'V312'), (437.59151942018076, 'V284'), (422.2044880149039, 'V288'), (265.3660076572749, 'V301'), (242.42351611869546, 'V292'), (206.587173675625, 'V131'), (159.41228364861615, 'V300'), (158.0075807024172, 'V287'), (130.7631632216822, 'V291'), (128.7257028057175, 'V122'), (126.06151462421106, 'V309'), (57.055851425447145, 'V129'), (52.70748303382973, 'V121'), (42.57680811160498, 'V310'), (23.592811961429494, 'V285'), (12.421571795102325, 'V307'), (11.497813825026581, 'V286'), (9.236325247603943, 'V317'), (6.839859709641704, 'V320'), (6.034911292101641, 'V294'), (5

In [185]:
#then we select the the best features

x_train_f=x_train_norm.loc[:,['V18','V17','ProductCD','V15','card3','card6','V16','V21','V22',
                        'V123','V302','V304','V303','V125','V283','V282','V87','V86','V94','V81',
                        'V93','V92','V85','V84','V111','V113','V112','V90','V91','V108','V114',
                        'V110','V116','V79','V33','V34','V74','V80','V73','V58','V31','V57',
                        'V32','V77','V72','V23','V78','V71','V63','V59','V64','V60','V24','V29',
                     'V69','V70','V30','D15','D10','D1','V67','V62','V56','V83','V109','V115',
                           'V124','V281']]

In [186]:
cross_val_score(sgd, x_train_f, y_train, cv=5, scoring="roc_auc")
#a lot better

array([0.75013302, 0.7432881 , 0.75178875, 0.75973005, 0.71728381])

In [187]:
y_train_pred=cross_val_predict(sgd,x_train_f,y_train,cv=5)
confusion_matrix(y_train,y_train_pred)
#ok, it's still bad

array([[305554,      0],
       [ 10973,      1]])

In [188]:
from sklearn.preprocessing import StandardScaler     #for standarization
x_train_fs=StandardScaler().fit_transform(x_train_f)

In [189]:
x_train_fs=pd.DataFrame(x_train_fs,columns=x_train_f.columns,index=x_train_f.index)

In [190]:
cross_val_score(sgd, x_train_fs, y_train, cv=5, scoring="roc_auc")

array([0.63961493, 0.6901977 , 0.73196622, 0.76950281, 0.63836396])

In [191]:
y_train_pred=cross_val_predict(sgd,x_train_fs,y_train,cv=5)
confusion_matrix(y_train,y_train_pred)  
#better

array([[305054,    500],
       [ 10132,    842]])

### Let's try a different model (random forest)

In [192]:
from sklearn.ensemble import RandomForestClassifier
forest_clf = RandomForestClassifier(random_state=42)

In [193]:
cross_val_score(forest_clf, x_train_f, y_train, cv=5,scoring='roc_auc')
#very good

array([0.84684   , 0.84961497, 0.848347  , 0.85384883, 0.85000098])

In [194]:
y_forest = cross_val_predict(forest_clf, x_train_f, y_train, cv=5)
confusion_matrix(y_train,y_forest)      #without standraization

array([[304303,   1251],
       [  7608,   3366]])

In [195]:
y_forest = cross_val_predict(forest_clf, x_train_fs, y_train, cv=5)
confusion_matrix(y_train,y_forest)    #the standraization doesn't do a lot 

array([[304298,   1256],
       [  7600,   3374]])

In [196]:
from sklearn.metrics import roc_auc_score
roc_auc_score(y_train, y_forest)  #we still can do much better

0.6516717078792688

In [197]:
#removing the outliers
q1=visa[visa.isFraud==0].quantile(0.05)
q2=visa[visa.isFraud==0].quantile(0.95)
iqr=q2-q1
mini=q1-1.5*iqr
maxi=q2+1.5*iqr
con=(visa[visa.isFraud==0]<=maxi)&(visa[visa.isFraud==0]>=mini)
con=con.all(axis=1)
visa_iqr=visa[visa.isFraud==0][con]  

In [198]:
visa_iqr.shape

(241960, 181)

In [199]:
visa_iqr.isFraud.value_counts()  #It seems that all 1 values are dropped :(

0.0    241960
Name: isFraud, dtype: int64

In [200]:
visa_new=pd.concat([visa_iqr,visa[visa.isFraud==1]])   
#making a dataframe of the 1 values with the data comes from the quantile (after removing outliers)

In [201]:
visa_new=visa_new.reset_index()
visa_new=visa_new.drop(columns=['index'])

In [202]:
visa_new.shape

(255677, 181)

In [203]:
#splitting the data 
for train_index,test_index in split.split(visa_new,visa_new['isFraud']):
    train_set_new=visa_new.loc[train_index]
    test_set_new=visa_new.loc[test_index]

In [204]:
x_train_new=train_set_new.drop(columns=['isFraud'])
y_train_new=train_set_new['isFraud']

In [205]:
x_test_new=test_set_new.drop(columns=['isFraud'])
y_test_new=test_set_new['isFraud']

In [206]:
#selecting the best features
x_test_newf=x_test_new.loc[:,['V18','V17','ProductCD','V15','card3','card6','V16','V21','V22','V123',
                        'V302','V304','V303','V125','V283','V282','V87','V86','V94','V81',
                        'V93','V92','V85','V84','V111','V113','V112','V90','V91','V108','V114',
                        'V110','V116','V79','V33','V34','V74','V80','V73','V58','V31','V57',
                        'V32','V77','V72','V23','V78','V71','V63','V59','V64','V60','V24','V29',
                        'V69','V70','V30','D15','D10','D1','V67','V62','V56','V83','V109','V115',
                           'V124','V281']]

In [207]:
x_train_newf=x_train_new.loc[:,['V18','V17','ProductCD','V15','card3','card6','V16','V21','V22','V123',
                        'V302','V304','V303','V125','V283','V282','V87','V86','V94','V81',
                        'V93','V92','V85','V84','V111','V113','V112','V90','V91','V108','V114',
                        'V110','V116','V79','V33','V34','V74','V80','V73','V58','V31','V57',
                        'V32','V77','V72','V23','V78','V71','V63','V59','V64','V60','V24','V29',
                        'V69','V70','V30','D15','D10','D1','V67','V62','V56','V83','V109','V115',
                           'V124','V281']]

In [208]:
x_train_newf.shape

(204541, 68)

In [209]:
y_train_new.shape

(204541,)

In [210]:
from imblearn.over_sampling import SMOTE       #for upsampling
sm = SMOTE(random_state=42)
x_train_res, y_train_res = sm.fit_sample(x_train_newf, y_train_new)

In [211]:
y_train_res.shape

(387134,)

In [212]:
y_train_res.value_counts()

1.0    193567
0.0    193567
Name: isFraud, dtype: int64

In [213]:
visa_super0=visa_new.loc[:,['isFraud','V18','V17','ProductCD','V15','card3','card6','V16','V21','V22','V123',
                        'V302','V304','V303','V125','V283','V282','V87','V86','V94','V81',
                        'V93','V92','V85','V84','V111','V113','V112','V90','V91','V108','V114',
                        'V110','V116','V79','V33','V34','V74','V80','V73','V58','V31','V57',
                        'V32','V77','V72','V23','V78','V71','V63','V59','V64','V60','V24','V29',
                        'V69','V70','V30','D15','D10','D1','V67','V62','V56','V83','V109','V115',
                           'V124','V281']]

In [214]:
y_forest = cross_val_predict(forest_clf, x_train_res, y_train_res, cv=5)
confusion_matrix(y_train_res,y_forest)
#much better

array([[186297,   7270],
       [ 37424, 156143]])

In [215]:
roc_auc_score(y_train_res,y_forest)
#great score but let's try to improve it

0.8845516022875799

# Let's start hyperparameter tuning using grid search

In [92]:
from sklearn.model_selection import GridSearchCV

param_grid = [

    {'n_estimators': [30,50,100,200], 'max_features': [2, 4, 6, 8]}
    
  ]

forest_clf = RandomForestClassifier(random_state=42)

grid_search = GridSearchCV(forest_clf, param_grid, cv=5,
                           scoring='roc_auc',
                           return_train_score=True)
grid_search.fit(x_train_res, y_train_res)

GridSearchCV(cv=5, estimator=RandomForestClassifier(random_state=42),
             param_grid=[{'max_features': [2, 4, 6, 8],
                          'n_estimators': [30, 50, 100, 200]}],
             return_train_score=True, scoring='roc_auc')

In [93]:
grid_search.best_params_     #the best params

{'max_features': 2, 'n_estimators': 200}

In [94]:
forest_clf = RandomForestClassifier(random_state=42,max_features=2,n_estimators=200)

In [95]:
y_forest = cross_val_predict(forest_clf, x_train_res, y_train_res, cv=5)
confusion_matrix(y_train_res,y_forest)
#great

array([[186312,   7255],
       [ 37427, 156140]])

In [96]:
roc_auc_score(y_train_res,y_forest)
#that's our best 

0.8845825993066999

In [97]:
forest_clf.fit(x_train_res,y_train_res)    #fitting the model
y_pre=forest_clf.predict(x_test_newf)
confusion_matrix(y_test_new,y_pre)

array([[46542,  1851],
       [  857,  1886]])

In [98]:
roc_auc_score(y_test_new,y_pre)

0.8246595111167491

### now we concat all the data we used

In [99]:
visa_mega0=pd.concat([x_train_res,x_test_newf])  

In [100]:
visa_mega1=pd.concat([y_train_res,y_test_new])

In [101]:
visa_mega=pd.concat([visa_mega0,visa_mega1],axis=1)

### Then we fit the model with the whole data 

In [103]:
forest_clf.fit(visa_mega.drop(columns='isFraud'),visa_mega['isFraud'])

RandomForestClassifier(max_features=2, n_estimators=200, random_state=42)

In [104]:
#The data we want to predict on
df2=pd.read_csv('/kaggle/input/data-science-circle-challenge/test.csv')

In [105]:
ID=df2['TransactionID']    #The first column of submission

### Preparing the test data

In [106]:
df2=df2.loc[:,['V18','V17','ProductCD','V15','card3','card6','V16','V21','V22','V123',
                        'V302','V304','V303','V125','V283','V282','V87','V86','V94','V81',
                        'V93','V92','V85','V84','V111','V113','V112','V90','V91','V108','V114',
                        'V110','V116','V79','V33','V34','V74','V80','V73','V58','V31','V57',
                        'V32','V77','V72','V23','V78','V71','V63','V59','V64','V60','V24','V29',
                        'V69','V70','V30','D15','D10','D1','V67','V62','V56','V83','V109','V115',
                           'V124','V281']]

In [107]:
df2.ProductCD=df2.ProductCD.astype('category')
df2.card6=df2.card6.astype('category')

In [108]:
df2.fillna(value={'ProductCD':df2.ProductCD.mode(),'card6':df2.card6.mode()},inplace=True)

In [109]:
for col in ['ProductCD','card6']:
    df2[col]=df2[col].cat.codes

In [110]:
df2_0=SimpleImputer(strategy='median').fit_transform(df2)
df2_0=pd.DataFrame(df2_0,columns=df2.columns,index=df2.index)

## Then predict with the probability

In [111]:
y_proba=forest_clf.predict_proba(df2_0)

In [117]:
y_f=pd.DataFrame(y_proba)

In [118]:
final=pd.concat([ID,y_f[1]],axis=1)  #preparing the final shape of submission

In [119]:
final=final.rename(columns={1: "isFraud"})

In [120]:
final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194879 entries, 0 to 194878
Data columns (total 2 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionID  194879 non-null  int64  
 1   isFraud        194879 non-null  float64
dtypes: float64(1), int64(1)
memory usage: 3.0 MB


## And we are ready 

In [116]:
final.to_csv('finaloX.csv',index=False)

# After submission we get about 82% roc_auc score 